src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -113,6 +115,254 @@ void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
     kh_destroy_inplace(uct_rdmacm_cm_cqs, &cm->cqs);
 }
 
+static ucs_status_t
+uct_rdmacm_cm_query_reserved_qpn_cap(struct ibv_context *verbs,
+                                     struct uct_rdmacm_cm_reserved_qpn_context *ctx)
+{
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_out)] = {};
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_in)]   = {};
+    void *cap;
+    int ret;
+
+    if (!mlx5dv_is_supported(verbs->device)) {
+        ctx->support = false;
+        ucs_debug("mlx5dv is not supported");
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    cap = UCT_IB_MLX5DV_ADDR_OF(query_hca_cap_out, out, capability);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, opcode,
+                      UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod,
+                      (UCT_IB_MLX5_CAP_GENERAL << 1) |
+                      UCT_IB_MLX5_HCA_CAP_OPMOD_GET_CUR);
+    ret = mlx5dv_devx_general_cmd(verbs, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        ctx->support = false;
+        ucs_debug("mlx5dv_devx_general_cmd(QUERY_HCA_CAP) failed: %m");
+        return UCS_ERR_UNSUPPORTED;
+    }
+    ctx->general_obj_types_caps = UCT_IB_MLX5DV_GET64(cmd_hca_cap, cap,
+                                                      general_obj_types);
+    if (!(ctx->general_obj_types_caps &
+        (1ULL << UCT_IB_MLX5_OBJ_TYPE_RESERVED_QPN))) {
+        ucs_debug("general_obj_types_caps: reserved qpn is not support");
+        ctx->support = false;
+        return UCS_ERR_UNSUPPORTED;
+    }
+    ctx->support = true;
+    return UCS_OK;
+#else
+    ctx->support = false;
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static ucs_status_t
+uct_rdmacm_cm_query_reserved_qpn_cap_2(struct ibv_context *verbs,
```

Comment:
why `_2`?

Reply:
@evgeny-leksikov The name of cap_2 comes from the PRM. In the PRM some caps are defined as hca_cap, some caps are defined as hca_cap_2.